File: legacy-go/internal/ui/ansi.go
package ui

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/rivo/tview"
)

// ansiToTview converts a pane's ANSI SGR sequences into tview colour tags. It
// re-states the FULL style on every SGR change, so an attribute (underline,
// dim, …) can never bleed past its reset — unlike tview.TranslateANSI, which
// emits nothing when the attribute set empties and so leaves the prior tag in
// effect. Non-SGR CSI sequences and OSC strings are dropped (never leaked as
// raw escape bytes), and literal "[tag]"-looking text is escaped.
func ansiToTview(s string) string {
	var out, text strings.Builder
	var fg, bg string // "" == terminal default
	flags := map[byte]bool{}

	flush := func() {
		if text.Len() > 0 {
			out.WriteString(tview.Escape(text.String()))
			text.Reset()
		}
	}
	emit := func() {
		flush()
		f, b := "-", "-"
		if fg != "" {
			f = fg
		}
		if bg != "" {
			b = bg
		}
		attr := ""
		for _, c := range []byte{'b', 'd', 'i', 'u', 'l', 'r', 's'} {
			if flags[c] {
				attr += string(c)
			}
		}
		if attr == "" {
			attr = "-"
		}
		fmt.Fprintf(&out, "[%s:%s:%s]", f, b, attr)
	}

	rs := []rune(s)
	for i := 0; i < len(rs); i++ {
		if rs[i] != 0x1b {
			text.WriteRune(rs[i])
			continue
		}
		if i+1 >= len(rs) {
			break
		}
		switch rs[i+1] {
		case '[': // CSI: params until a final byte 0x40–0x7e
			j := i + 2
			for j < len(rs) && (rs[j] < 0x40 || rs[j] > 0x7e) {
				j++
			}
			if j >= len(rs) {
				i = len(rs)
				break
			}
			if rs[j] == 'm' {
				applySGR(string(rs[i+2:j]), &fg, &bg, flags)
				emit()
			}
			i = j
		case ']': // OSC: until BEL or ST (ESC \)
			j := i + 2
			for j < len(rs) {
				if rs[j] == 0x07 {
					break
				}
				if rs[j] == 0x1b && j+1 < len(rs) && rs[j+1] == '\\' {
					j++
					break
				}
				j++
			}
			i = j
		default:
			i++ // consume ESC + the next byte
		}
	}
	flush()
	return out.String()
}

var ansi16 = []string{
	"black", "maroon", "green", "olive", "navy", "purple", "teal", "silver",
	"gray", "red", "lime", "yellow", "blue", "fuchsia", "aqua", "white",
}

func basicColor(n int) string {
	if n < 0 || n > 15 {
		return ""
	}
	return ansi16[n]
}

func applySGR(params string, fg, bg *string, flags map[byte]bool) {
	if params == "" {
		params = "0"
	}
	fields := strings.Split(params, ";")
	for i := 0; i < len(fields); i++ {
		switch f := fields[i]; f {
		case "0", "":
			*fg, *bg = "", ""
			for k := range flags {
				delete(flags, k)
			}
		case "1", "01":
			flags['b'] = true
		case "2", "02":
			flags['d'] = true
		case "3", "03":
			flags['i'] = true
		case "4", "04":
			flags['u'] = true
		case "5", "05":
			flags['l'] = true
		case "7", "07":
			flags['r'] = true
		case "9", "09":
			flags['s'] = true
		case "22":
			delete(flags, 'b')
			delete(flags, 'd')
		case "23":
			delete(flags, 'i')
		case "24":
			delete(flags, 'u')
		case "25":
			delete(flags, 'l')
		case "27":
			delete(flags, 'r')
		case "29":
			delete(flags, 's')
		case "39":
			*fg = ""
		case "49":
			*bg = ""
		case "38", "48":
			color, consumed := extendedColor(fields, i)
			if color != "" {
				if f == "38" {
					*fg = color
				} else {
					*bg = color
				}
			}
			i += consumed
		case "58": // underline colour — ignore, but consume its sub-params
			_, consumed := extendedColor(fields, i)
			i += consumed
		default:
			if n, err := strconv.Atoi(f); err == nil {
				switch {
				case n >= 30 && n <= 37:
					*fg = basicColor(n - 30)
				case n >= 40 && n <= 47:
					*bg = basicColor(n - 40)
				case n >= 90 && n <= 97:
					*fg = basicColor(n - 82)
				case n >= 100 && n <= 107:
					*bg = basicColor(n - 92)
				}
			}
			// anything else (59, colon-forms like "4:3", …) is ignored.
		}
	}
}

// extendedColor parses a 38/48/58 colour starting at fields[i] and returns the
// tview colour plus how many EXTRA fields it consumed.
func extendedColor(fields []string, i int) (string, int) {
	if i+1 >= len(fields) {
		return "", 0
	}
	switch fields[i+1] {
	case "5": // 8-bit
		if i+2 >= len(fields) {
			return "", 1
		}
		n, _ := strconv.Atoi(fields[i+2])
		switch {
		case n <= 15:
			return basicColor(n), 2
		case n <= 231:
			r, g, b := (n-16)/36, ((n-16)/6)%6, (n-16)%6
			return fmt.Sprintf("#%02x%02x%02x", 255*r/5, 255*g/5, 255*b/5), 2
		case n <= 255:
			grey := 255 * (n - 232) / 23
			return fmt.Sprintf("#%02x%02x%02x", grey, grey, grey), 2
		}
		return "", 2
	case "2": // 24-bit
		if i+4 >= len(fields) {
			return "", len(fields) - i - 1
		}
		r, _ := strconv.Atoi(fields[i+2])
		g, _ := strconv.Atoi(fields[i+3])
		b, _ := strconv.Atoi(fields[i+4])
		return fmt.Sprintf("#%02x%02x%02x", r, g, b), 4
	}
	return "", 0
}

File: src/ui/ansi.rs
//! Converts a pane's ANSI SGR sequences into a ratatui [`Text`] of styled
//! [`Span`]s. Each span carries its FULL resolved style, so an attribute
//! (underline, dim, …) can never bleed past its reset — every span is rendered
//! independently. Non-SGR CSI sequences and OSC strings are dropped (never leaked
//! as raw escape bytes). Unlike the tview port this does not interpret or escape
//! markup, because ratatui spans hold literal text.

use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span, Text};

/// The accumulating SGR state: a current fg/bg and modifier set.
#[derive(Clone, Default)]
struct SgrState {
    fg: Option<Color>,
    bg: Option<Color>,
    mods: Modifier,
}

impl SgrState {
    fn style(&self) -> Style {
        let mut s = Style::default();
        if let Some(c) = self.fg {
            s = s.fg(c);
        }
        if let Some(c) = self.bg {
            s = s.bg(c);
        }
        s.add_modifier(self.mods)
    }
}

/// Converts an ANSI string into a styled [`Text`]. `\n` splits lines; a trailing
/// newline does not produce a spurious empty line.
pub fn ansi_to_text(s: &str) -> Text<'static> {
    let chars: Vec<char> = s.chars().collect();
    let len = chars.len();
    let mut lines: Vec<Line<'static>> = Vec::new();
    let mut spans: Vec<Span<'static>> = Vec::new();
    let mut cur = String::new();
    let mut state = SgrState::default();

    let mut i = 0;
    while i < len {
        let c = chars[i];
        match c {
            '\n' => {
                flush_span(&mut spans, &mut cur, &state);
                lines.push(Line::from(std::mem::take(&mut spans)));
                i += 1;
            }
            '\r' => {
                i += 1; // drop a bare CR
            }
            '\u{1b}' => {
                if i + 1 >= len {
                    break;
                }
                match chars[i + 1] {
                    '[' => {
                        // CSI: params until a final byte 0x40–0x7e.
                        let mut j = i + 2;
                        while j < len && !('\u{40}'..='\u{7e}').contains(&chars[j]) {
                            j += 1;
                        }
                        if j >= len {
                            break; // unterminated CSI — drop the rest
                        }
                        if chars[j] == 'm' {
                            // Flush text accumulated under the PRIOR style first.
                            flush_span(&mut spans, &mut cur, &state);
                            let params: String = chars[i + 2..j].iter().collect();
                            apply_sgr(&params, &mut state);
                        }
                        i = j + 1;
                    }
                    ']' => {
                        // OSC: until BEL or ST (ESC \).
                        let mut j = i + 2;
                        while j < len {
                            if chars[j] == '\u{07}' {
                                break;
                            }
                            if chars[j] == '\u{1b}' && j + 1 < len && chars[j + 1] == '\\' {
                                j += 1;
                                break;
                            }
                            j += 1;
                        }
                        i = j + 1;
                    }
                    _ => {
                        i += 2; // consume ESC + the next byte
                    }
                }
            }
            _ => {
                cur.push(c);
                i += 1;
            }
        }
    }

    flush_span(&mut spans, &mut cur, &state);
    if !spans.is_empty() {
        lines.push(Line::from(spans));
    }
    Text::from(lines)
}

fn flush_span(spans: &mut Vec<Span<'static>>, cur: &mut String, state: &SgrState) {
    if !cur.is_empty() {
        spans.push(Span::styled(std::mem::take(cur), state.style()));
    }
}

fn basic_color(n: i32) -> Option<Color> {
    if (0..=15).contains(&n) {
        Some(Color::Indexed(n as u8))
    } else {
        None
    }
}

fn apply_sgr(params: &str, state: &mut SgrState) {
    let params = if params.is_empty() { "0" } else { params };
    let fields: Vec<&str> = params.split(';').collect();
    let mut i = 0;
    while i < fields.len() {
        match fields[i] {
            "0" | "" => {
                state.fg = None;
                state.bg = None;
                state.mods = Modifier::empty();
            }
            "1" | "01" => state.mods.insert(Modifier::BOLD),
            "2" | "02" => state.mods.insert(Modifier::DIM),
            "3" | "03" => state.mods.insert(Modifier::ITALIC),
            "4" | "04" => state.mods.insert(Modifier::UNDERLINED),
            "5" | "05" => state.mods.insert(Modifier::SLOW_BLINK),
            "7" | "07" => state.mods.insert(Modifier::REVERSED),
            "9" | "09" => state.mods.insert(Modifier::CROSSED_OUT),
            "22" => state.mods.remove(Modifier::BOLD | Modifier::DIM),
            "23" => state.mods.remove(Modifier::ITALIC),
            "24" => state.mods.remove(Modifier::UNDERLINED),
            "25" => state.mods.remove(Modifier::SLOW_BLINK),
            "27" => state.mods.remove(Modifier::REVERSED),
            "29" => state.mods.remove(Modifier::CROSSED_OUT),
            "39" => state.fg = None,
            "49" => state.bg = None,
            "38" | "48" => {
                let (color, consumed) = extended_color(&fields, i);
                if let Some(c) = color {
                    if fields[i] == "38" {
                        state.fg = Some(c);
                    } else {
                        state.bg = Some(c);
                    }
                }
                i += consumed;
            }
            "58" => {
                // underline colour — ignore, but consume its sub-params.
                let (_c, consumed) = extended_color(&fields, i);
                i += consumed;
            }
            other => {
                if let Ok(n) = other.parse::<i32>() {
                    match n {
                        30..=37 => state.fg = basic_color(n - 30),
                        40..=47 => state.bg = basic_color(n - 40),
                        90..=97 => state.fg = basic_color(n - 82),
                        100..=107 => state.bg = basic_color(n - 92),
                        _ => {}
                    }
                }
                // anything else (59, colon-forms like "4:3", …) is ignored.
            }
        }
        i += 1;
    }
}

/// Parses a `38`/`48`/`58` colour starting at `fields[i]` and returns the colour
/// plus how many EXTRA fields it consumed.
fn extended_color(fields: &[&str], i: usize) -> (Option<Color>, usize) {
    if i + 1 >= fields.len() {
        return (None, 0);
    }
    match fields[i + 1] {
        "5" => {
            // 8-bit: any palette index 0–255 maps to Indexed (the terminal's own
            // theme decides the actual colour).
            if i + 2 >= fields.len() {
                return (None, 1);
            }
            match fields[i + 2].parse::<u16>() {
                Ok(n) if n <= 255 => (Some(Color::Indexed(n as u8)), 2),
                _ => (None, 2),
            }
        }
        "2" => {
            // 24-bit truecolor.
            if i + 4 >= fields.len() {
                return (None, fields.len() - i - 1);
            }
            let r = fields[i + 2].parse::<u8>().unwrap_or(0);
            let g = fields[i + 3].parse::<u8>().unwrap_or(0);
            let b = fields[i + 4].parse::<u8>().unwrap_or(0);
            (Some(Color::Rgb(r, g, b)), 4)
        }
        _ => (None, 0),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Flattens a Text into (text, style) span tuples.
    fn spans(t: &Text<'static>) -> Vec<(String, Style)> {
        let mut out = Vec::new();
        for line in &t.lines {
            for span in &line.spans {
                out.push((span.content.to_string(), span.style));
            }
        }
        out
    }

    fn find(spans: &[(String, Style)], text: &str) -> Style {
        spans
            .iter()
            .find(|(c, _)| c == text)
            .unwrap_or_else(|| panic!("no span with text {text:?} in {spans:?}"))
            .1
    }

    fn flat(t: &Text<'static>) -> String {
        spans(t).into_iter().map(|(c, _)| c).collect()
    }

    #[test]
    fn no_underline_bleed() {
        let t = ansi_to_text("\x1b[4mUNDER\x1b[24mNORMAL");
        let sp = spans(&t);
        assert!(find(&sp, "UNDER")
            .add_modifier
            .contains(Modifier::UNDERLINED));
        assert!(!find(&sp, "NORMAL")
            .add_modifier
            .contains(Modifier::UNDERLINED));
    }

    #[test]
    fn reset_clears_attributes() {
        let t = ansi_to_text("\x1b[4mU\x1b[0m end");
        let sp = spans(&t);
        assert_eq!(find(&sp, " end"), Style::default());
    }

    #[test]
    fn truecolor_then_underline() {
        let t = ansi_to_text("\x1b[38;2;255;0;0;4mX");
        let style = find(&spans(&t), "X");
        assert_eq!(style.fg, Some(Color::Rgb(255, 0, 0)));
        assert!(style.add_modifier.contains(Modifier::UNDERLINED));
    }

    #[test]
    fn underline_color_ignored_not_misparsed() {
        // 58;2;r;g;b (underline colour) must be consumed, not parsed as 2=dim.
        let t = ansi_to_text("\x1b[4;58;2;0;255;0mX");
        let style = find(&spans(&t), "X");
        assert!(style.add_modifier.contains(Modifier::UNDERLINED));
        assert!(!style.add_modifier.contains(Modifier::DIM));
    }

    #[test]
    fn basic_and_256_colors() {
        let t = ansi_to_text("\x1b[31mR");
        let sp = spans(&t);
        assert_eq!(find(&sp, "R").fg, Some(Color::Indexed(1)));
        let t = ansi_to_text("\x1b[38;5;9mR");
        let sp = spans(&t);
        assert_eq!(find(&sp, "R").fg, Some(Color::Indexed(9)));
    }

    #[test]
    fn bright_foreground_maps_to_high_palette() {
        // 91 = bright red = palette index 9.
        let t = ansi_to_text("\x1b[91mR");
        let sp = spans(&t);
        assert_eq!(find(&sp, "R").fg, Some(Color::Indexed(9)));
    }

    #[test]
    fn drops_unknown_sequences() {
        // An OSC hyperlink and a non-SGR CSI must be dropped, not leaked.
        let t = ansi_to_text("a\x1b]8;;http://x\x07b\x1b[2Kc");
        let s = flat(&t);
        assert!(!s.contains('\u{1b}'), "no raw escape may leak: {s:?}");
        for w in ["a", "b", "c"] {
            assert!(s.contains(w), "text {w:?} lost: {s:?}");
        }
    }

    #[test]
    fn newlines_split_lines_no_trailing_empty() {
        let t = ansi_to_text("one\ntwo");
        assert_eq!(t.lines.len(), 2);
        let t2 = ansi_to_text("one\n");
        assert_eq!(t2.lines.len(), 1);
    }
}
